### ISSUE TITLE
sslcertkey attribute of lbvserver not refreshing properly when CA certificates are bound

### ISSUE THREAD

**Issue by jegelandsdal** [citrix/terraform-provider-citrixadc#81]
**Describe the bug**
Resource citrixadc_lbvserver does not refresh sslcertkey attribute properly when CA certificates are bound to virtual server.

**To reproduce**

Attempt to update a citrixadc_lbvserver resource with an sslcertkey attribute, while the virtual server has a CA certificate bound.

Terraform configuration:

```
variable "netscaler" {
    description = "Map for Nitro API access information.  It has attributes hostname, username, and password."
    type = object({hostname = string, username = string, password = string})
}

variable "virtualserver_ip" {
    description = "virtual server IP address"
}

variable "server_cert" {
    description = "Server certificate name"
}

variable "ca_chain" {
    description = "Map with CA certificate chain names.  It has attributes int and root."
    type = object({int = string, root = string})
}

locals {
        endpoint = "https://${var.netscaler.hostname}/"
}

provider "citrixadc" {
    endpoint = local.endpoint
    username = var.netscaler.username
    password = var.netscaler.password
    insecure_skip_verify = true
}

resource "citrixadc_lbvserver" "ssl_virtualserver" {
        name = "certkey-error-test"
        ipv46 = var.virtualserver_ip
        port = 443
        servicetype = "SSL"
        persistencetype = "SSLSESSION"
        sslcertkey = var.server_cert

        provisioner "local-exec" {
        command = <<HERE
sshpass -p ${var.netscaler.password} ssh -o StrictHostKeyChecking=no ${var.netscaler.username}@${var.netscaler.hostname} <<DOC
bind ssl vserver ${self.name} -certkeyName ${var.ca_chain.int} -CA -ocspCheck Optional
bind ssl vserver ${self.name} -certkeyName ${var.ca_chain.root} -CA -ocspCheck Optional
DOC
        HERE
        }
}

```
**Expected behavior**
CA certificates bound to a virtual server will have no bearing on the sslcertkey attribute within a citrixadc_lbvserver resource.

**Logs**

Error output below.  Sanitized trace logs can be obtained as well if necessary.

```
terraform@machine:~/CitrixADC_Certkey_Error$ terraform apply --var-file input.tfvars 

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # citrixadc_lbvserver.ssl_virtualserver will be created
  + resource "citrixadc_lbvserver" "ssl_virtualserver" {
      + appflowlog                         = (known after apply)
      + authentication                     = (known after apply)
      + authenticationhost                 = (known after apply)
      + authn401                           = (known after apply)
      + authnprofile                       = (known after apply)
      + authnvsname                        = (known after apply)
      + backuplbmethod                     = (known after apply)
      + backuppersistencetimeout           = (known after apply)
      + backupvserver                      = (known after apply)
      + bypassaaaa                         = (known after apply)
      + cacheable                          = (known after apply)
      + clttimeout                         = (known after apply)
      + comment                            = (known after apply)
      + connfailover                       = (known after apply)
      + cookiename                         = (known after apply)
      + datalength                         = (known after apply)
      + dataoffset                         = (known after apply)
      + dbprofilename                      = (known after apply)
      + dbslb                              = (known after apply)
      + disableprimaryondown               = (known after apply)
      + dns64                              = (known after apply)
      + dnsprofilename                     = (known after apply)
      + downstateflush                     = (known after apply)
      + hashlength                         = (known after apply)
      + healththreshold                    = (known after apply)
      + httpprofilename                    = (known after apply)
      + httpsredirecturl                   = (known after apply)
      + icmpvsrresponse                    = (known after apply)
      + id                                 = (known after apply)
      + insertvserveripport                = (known after apply)
      + ipmask                             = (known after apply)
      + ippattern                          = (known after apply)
      + ipset                              = (known after apply)
      + ipv46                              = "10.1.1.20"
      + l2conn                             = (known after apply)
      + lbmethod                           = (known after apply)
      + lbprofilename                      = (known after apply)
      + listenpolicy                       = (known after apply)
      + listenpriority                     = (known after apply)
      + m                                  = (known after apply)
      + macmoderetainvlan                  = (known after apply)
      + maxautoscalemembers                = (known after apply)
      + minautoscalemembers                = (known after apply)
      + mssqlserverversion                 = (known after apply)
      + mysqlcharacterset                  = (known after apply)
      + mysqlprotocolversion               = (known after apply)
      + mysqlservercapabilities            = (known after apply)
      + mysqlserverversion                 = (known after apply)
      + name                               = "certkey-error-test"
      + netmask                            = (known after apply)
      + netprofile                         = (known after apply)
      + newname                            = (known after apply)
      + newservicerequest                  = (known after apply)
      + newservicerequestincrementinterval = (known after apply)
      + newservicerequestunit              = (known after apply)
      + oracleserverversion                = (known after apply)
      + persistencebackup                  = (known after apply)
      + persistencetype                    = "SSLSESSION"
      + persistmask                        = (known after apply)
      + port                               = 443
      + pq                                 = (known after apply)
      + processlocal                       = (known after apply)
      + push                               = (known after apply)
      + pushlabel                          = (known after apply)
      + pushmulticlients                   = (known after apply)
      + pushvserver                        = (known after apply)
      + range                              = (known after apply)
      + recursionavailable                 = (known after apply)
      + redirectfromport                   = (known after apply)
      + redirectportrewrite                = (known after apply)
      + redirurl                           = (known after apply)
      + redirurlflags                      = (known after apply)
      + resrule                            = (known after apply)
      + retainconnectionsoncluster         = (known after apply)
      + rhistate                           = (known after apply)
      + rtspnat                            = (known after apply)
      + rule                               = (known after apply)
      + sc                                 = (known after apply)
      + servicename                        = (known after apply)
      + servicetype                        = "SSL"
      + sessionless                        = (known after apply)
      + skippersistency                    = (known after apply)
      + sobackupaction                     = (known after apply)
      + somethod                           = (known after apply)
      + sopersistence                      = (known after apply)
      + sopersistencetimeout               = (known after apply)
      + sothreshold                        = (known after apply)
      + sslcertkey                         = "serverCert"
      + state                              = (known after apply)
      + tcpprofilename                     = (known after apply)
      + td                                 = (known after apply)
      + timeout                            = (known after apply)
      + tosid                              = (known after apply)
      + trofspersistence                   = (known after apply)
      + v6netmasklen                       = (known after apply)
      + v6persistmasklen                   = (known after apply)
      + vipheader                          = (known after apply)
      + weight                             = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

citrixadc_lbvserver.ssl_virtualserver: Creating...
citrixadc_lbvserver.ssl_virtualserver: Provisioning with 'local-exec'...
citrixadc_lbvserver.ssl_virtualserver (local-exec): Executing: ["/bin/sh" "-c" "sshpass -p password ssh -o StrictHostKeyChecking=no username@ipaddress <<DOC\r\nbind ssl vserver certkey-error-test -certkeyName ca_int -CA -ocspCheck Optional\r\nbind ssl vserver certkey-error-test -certkeyName ca_root -CA -ocspCheck Optional\r\nDOC\r\n"]
citrixadc_lbvserver.ssl_virtualserver (local-exec): Pseudo-terminal will not be allocated because stdin is not a terminal.

citrixadc_lbvserver.ssl_virtualserver (local-exec):  Done
citrixadc_lbvserver.ssl_virtualserver (local-exec):  Done
citrixadc_lbvserver.ssl_virtualserver (local-exec):  Done
citrixadc_lbvserver.ssl_virtualserver (local-exec): ^D

citrixadc_lbvserver.ssl_virtualserver (local-exec): Fatal Exception!: EOF
citrixadc_lbvserver.ssl_virtualserver (local-exec): Bye!
citrixadc_lbvserver.ssl_virtualserver: Creation complete after 1s [id=certkey-error-test]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
terraform@machine:~/CitrixADC_Certkey_Error$ terraform apply --var-file input.tfvars 
citrixadc_lbvserver.ssl_virtualserver: Refreshing state... [id=certkey-error-test]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # citrixadc_lbvserver.ssl_virtualserver will be updated in-place
  ~ resource "citrixadc_lbvserver" "ssl_virtualserver" {
        appflowlog                         = "ENABLED"
        authentication                     = "OFF"
        authn401                           = "OFF"
        backuplbmethod                     = "ROUNDROBIN"
        backuppersistencetimeout           = 2
        bypassaaaa                         = "NO"
        cacheable                          = "NO"
      ~ ciphers                            = [
          - "DEFAULT",
        ]
        connfailover                       = "DISABLED"
        disableprimaryondown               = "DISABLED"
        dns64                              = "DISABLED"
        downstateflush                     = "ENABLED"
        hashlength                         = 0
        icmpvsrresponse                    = "PASSIVE"
        id                                 = "certkey-error-test"
        insertvserveripport                = "OFF"
        ipmask                             = "*"
        ippattern                          = "0.0.0.0"
        ipv46                              = "10.1.1.20"
        l2conn                             = "OFF"
        lbmethod                           = "LEASTCONNECTION"
        listenpolicy                       = "NONE"
        listenpriority                     = 0
        m                                  = "IP"
        macmoderetainvlan                  = "DISABLED"
        mysqlcharacterset                  = 0
        mysqlprotocolversion               = 0
        mysqlservercapabilities            = 0
        name                               = "certkey-error-test"
        newservicerequest                  = 0
        newservicerequestincrementinterval = 0
        newservicerequestunit              = "PER_SECOND"
        persistencebackup                  = "NONE"
        persistencetype                    = "SSLSESSION"
        persistmask                        = "255.255.255.255"
        port                               = 443
        processlocal                       = "DISABLED"
        push                               = "DISABLED"
        pushlabel                          = "none"
        pushmulticlients                   = "NO"
        redirectfromport                   = 0
        redirectportrewrite                = "DISABLED"
        redirurlflags                      = false
        retainconnectionsoncluster         = "NO"
        rhistate                           = "PASSIVE"
        rtspnat                            = "OFF"
        servicetype                        = "SSL"
        sessionless                        = "DISABLED"
        skippersistency                    = "None"
        snisslcertkeys                     = []
        somethod                           = "NONE"
        sopersistence                      = "DISABLED"
        sothreshold                        = 0
      ~ sslcertkey                         = "ca_root" -> "serverCert"
        timeout                            = 2
        tosid                              = 0
        trofspersistence                   = "ENABLED"
        v6netmasklen                       = 0
        weight                             = 0
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

citrixadc_lbvserver.ssl_virtualserver: Modifying... [id=certkey-error-test]

Error: [ERROR] netscaler-provider: Error unbinding sslcertkey from lbvserver ca_root

  on main.tf line 30, in resource "citrixadc_lbvserver" "ssl_virtualserver":
  30: resource "citrixadc_lbvserver" "ssl_virtualserver" {

```

**Environment**
OS: Ubuntu 16.04.6 LTS
Terraform version: 0.12.23
CitrixADC version: 0.12.11

**Additional context**
It appears that the ability to bind CA certificates to a virtual server, for use with client authentication, has not been implemented in the provider yet.  As a workaround, it is necessary to perform this binding using a local-exec provisioner, which uses SSH to configure the Netscaler and perform the binding.  This works fine upon first apply.  However, subsequent applies incorrectly determine that the CA certificate is bound as the server certificate, when the provider performs a refresh.  This results in the provider attempting to update the sslcertkey attribute of the citrixadc_lbvserver resource.  This ultimately fails, as the provider is unable to unbind the CA certificate from the virtual server, in order to bind the server certificate.

The configuration does configure the Netscaler appropriately, and the server certificate and two CA certificates are bound appropriately to the virtual server, following an apply.  However, the error on subsequent applies is preventing the Terraform run from completing, and therefore not pushing all intentional configuration changes.  For instance, future changes to the virtual server do not succeed with consistency.

**Comment by giorgos-nikolopoulos** [comment]
I have verified the error you are seeing.

I think the best solution for this would be to implement a separate resource for sslvserver and sslcertkey bindings.
This will give the user the ability to fine tune the sslcertkey bindings.

We will take this up when time permits.

**Comment by giorgos-nikolopoulos** [comment]
Fixed with #137

There is now an explicit resource for binding a sslvserver and a certkey.

Using this instead of the lbvserver or csvserver attributes should allow finer control over the binding properties.
